vimPlugins.avante-nvim: Init at 2024-09-15

ttrei opened the pull request:
## Description of changes

New package: https://github.com/yetone/avante.nvim
Neovim plugin designed to emulate the behaviour of the [Cursor](https://www.cursor.com/) AI IDE.

Requested here: https://github.com/NixOS/nixpkgs/issues/339382

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ttrei commented:
0.0.2 seems broken, claude sonnet always complains about missing context.

aarnphm commented on pkgs/applications/editors/vim/plugins/avante-nvim/default.nix:
```diff
@@ -0,0 +1,65 @@
+  # The plugin expects the dynamic libraries to be under build/
+  postInstall = ''
+    mkdir -p $out/build
+    ln -s ${avante-lib}/*.so $out/build
```
instead of ln I think you can just copy instead.

ttrei replied:
Is there a downside to using symlinks here?
I used them to avoid duplicating the *.so files in nix store, and I saw that quite a lot of other plugins also use symlinks.

caarlos0 commented on pkgs/applications/editors/vim/plugins/avante-nvim/default.nix:
```diff
@@ -0,0 +1,65 @@
+  src = fetchFromGitHub {
+    owner = "yetone";
+    repo = "avante.nvim";
+    rev = "refs/tags/v${version}";
```
is it worth using the last commit like the other vim plugins do?

ttrei replied:
Indeed, that will help since the latest release is broken.

caarlos0 commented on pkgs/applications/editors/vim/plugins/overrides.nix:
```diff
@@ -152,6 +152,18 @@
+      nvim-web-devicons
+      nui-nvim
+      plenary-nvim
+      render-markdown
```
i think both render-markdown and img-clip-nvim are optional dependencies, no?

ttrei replied:
According to
https://github.com/yetone/avante.nvim/blob/7532e590d28c6fffe88aa17ba6689c4ec8fec8ba/README.md
these should be optional:
```
copilot-lua
img-clip-nvim
nvim-web-devicons
render-markdown
```

But i get an error on nvim startup if img-clip-nvim is not installed.

I guess we can remove the other three.
We don't have a mechanism for specifying optional dependencies, right?

vinnymeller commented:
don't have any extra feedback on the diff but this is working perfectly for me so far :+1: thanks!

vinnymeller commented on pkgs/applications/editors/vim/plugins/avante-nvim/default.nix:
```diff
@@ -0,0 +1,65 @@
+  rustPlatform,
+  fetchFromGitHub,
+  pkg-config,
+  openssl,
```
```suggestion
  openssl,
  Security,
  stdenv,
```

vinnymeller commented on pkgs/applications/editors/vim/plugins/avante-nvim/default.nix:
```diff
@@ -0,0 +1,65 @@
+      pkg-config
+      openssl
+    ];
+
```
```suggestion
    buildInputs = lib.optionals stdenv.isDarwin [ Security ];
    
```

vinnymeller commented on pkgs/applications/editors/vim/plugins/avante-nvim/default.nix:
```diff
@@ -0,0 +1,65 @@
+
+    installPhase = ''
+      mkdir -p $out
+      cp ./build/*.so $out/
```
```suggestion
      cp ./build/*.{dylib,so} $out/
```

vinnymeller commented on pkgs/applications/editors/vim/plugins/avante-nvim/default.nix:
```diff
@@ -0,0 +1,65 @@
+  # The plugin expects the dynamic libraries to be under build/
+  postInstall = ''
+    mkdir -p $out/build
+    ln -s ${avante-lib}/*.so $out/build
```
```suggestion
    ln -s ${avante-lib}/*.{dylib,so} $out/build
```

vinnymeller commented on pkgs/applications/editors/vim/plugins/overrides.nix:
```diff
@@ -152,6 +152,18 @@
     dependencies = with super; [ plenary-nvim ];
   };
 
+  avante-nvim = (callPackage ./avante-nvim { }).overrideAttrs {
```
```suggestion
  avante-nvim = (callPackage ./avante-nvim { inherit (darwin.apple_sdk.frameworks) Security; }).overrideAttrs {
```

vinnymeller reviewed (changes requested):
Comments show changes needed to build + work properly on macos

Commit from my fork for reference in case I typod something here: https://github.com/NixOS/nixpkgs/commit/bd270a4ed0fc70a8bc320507f2f1dd6b24dd02f5

ttrei commented:
@vinnymeller, thanks!
I incorporated your diff with a small change. Please verify that it works for you.

vinnymeller commented:
seems to be working fine :slightly_smiling_face:

vinnymeller reviewed (approved):
code looks good from my pov

i imagine they'll want the commits squashed though

ttrei commented:
squashed and rebased onto latest master.

caarlos0 commented:
just tried this, seems to work!

PS: I think img-clip can be removed... avante only complains about it not being there once and then just work normally

ttrei commented:
> PS: I think img-clip can be removed... avante only complains about it not being there once and then just work normally

yeah, i was wrong, it really is optional. i just had `require('img-clip').setup ({})` in my config :roll_eyes:

caarlos0 commented:
I'm a bit confused... it seemed like it worked for a while, but now I'm getting this: https://github.com/yetone/avante.nvim/issues/544

haven't really changed anything, so unsure of why this is happening

ttrei commented:
@caarlos0, please check if the libraries are present in nix store:
```
$ nvim --headless -c ':scriptnames' -c 'q!' 2>&1 | grep avante
 34: /nix/store/xr8j7a7y5r5skicnwbi9pqbi70zmgaap-vimplugin-avante.nvim-2024-09-08/plugin/avante.lua
$ ls -l /nix/store/xr8j7a7y5r5skicnwbi9pqbi70zmgaap-vimplugin-avante.nvim-2024-09-08/build
total 8.0K
lrwxrwxrwx 3 root root 90 Jan  1  1970 avante_templates.so -> /nix/store/yf7x6swkfz2pyl6lpf2kpig97dablz03-avante-nvim-lib-2024-09-08/avante_templates.so
lrwxrwxrwx 3 root root 91 Jan  1  1970 avante_tokenizers.so -> /nix/store/yf7x6swkfz2pyl6lpf2kpig97dablz03-avante-nvim-lib-2024-09-08/avante_tokenizers.so
```

caarlos0 commented:
yep, no build dir in there... weird, trying to figure it out

caarlos0 commented:
okay, dunno how it worked before, but apparently you also need to `require('avante_lib').load()`

added that to my config, now it works...

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4578

aarnphm commented on pkgs/applications/editors/vim/plugins/avante-nvim/default.nix:
```diff
@@ -0,0 +1,69 @@
+    cargoLock = {
+      lockFile = ./Cargo.lock;
+      outputHashes = {
+        "mlua-0.10.0-beta.1" = "sha256-ZEZFATVldwj0pmlmi0s5VT0eABA15qKhgjmganrhGBY=";
```
hmm btw we also use minijinja crates

ttrei replied:
it's specified in Cargo.lock

aarnphm commented on pkgs/applications/editors/vim/plugins/avante-nvim/default.nix:
```diff
@@ -0,0 +1,69 @@
+
+    installPhase = ''
+      mkdir -p $out
+      cp ./build/*.{dylib,so} $out/
```
hmm not sure about convention, but I usually determine extension statically through `stdlib.isDarwin` and `stdlib.isLinux` and reference it in here instead.

ttrei replied:
fixed

aarnphm commented on pkgs/applications/editors/vim/plugins/overrides.nix:
```diff
@@ -154,6 +154,14 @@
 
+  avante-nvim = (callPackage ./avante-nvim { }).overrideAttrs {
+    dependencies = with self; [
+      dressing-nvim
```
can you also add nvim-treesitter?

we will merge https://github.com/yetone/avante.nvim/pull/496 soon which requires treesitter.

ttrei replied:
let's add it when it's actually required

aarnphm reviewed (approved):
This LGTM on the first pass. Since we are adding a few more rust crates it would need some updates with that as well (but overall it is a good starting points, thanks a lot for your work).

ttrei committed:
maintainers: add ttrei

ttrei committed:
vimPlugins.avante-nvim: Init at 2024-09-15

ttrei commented:
Incorporated some improvements from the duplicate pr https://github.com/NixOS/nixpkgs/pull/341927
- buildRustPackage without custom build/install step
- nvimRequireCheck (+ fixed a bug in the check hook)

It was also suggested to add updateScript (https://github.com/NixOS/nixpkgs/pull/341927#pullrequestreview-2305301232)
Experimented a bit, seemed that it doesn't actually download Cargo.lock.

Added myself as maintainer.

Updated to the latest commit from main.

@vinnymeller, @aarnphm i'm requesting additional review, as these are not trivial changes:
[compare/5f07b177bf436029099f25a733c0a07b9faa1034..b9a52e994b794fdab8965ac488149249cefe3bbf](https://github.com/NixOS/nixpkgs/compare/5f07b177bf436029099f25a733c0a07b9faa1034..b9a52e994b794fdab8965ac488149249cefe3bbf)

vinnymeller reviewed (approved):
tested current commit + changes lgtm

aarnphm commented:
also feel free to add me as maintainer. I can keep an eye on this

aarnphm commented on pkgs/applications/editors/vim/plugins/neovim-require-check-hook.sh:
```diff
@@ -10,7 +10,7 @@ neovimRequireCheckHook () {
 		# editorconfig-checker-disable
         export HOME="$TMPDIR"
         @nvimBinary@ -es --headless -n -u NONE -i NONE --clean -V1 \
-            --cmd "set rtp+=$out,${dependencies/ /,}" \
```
hmm can you explain why we need to do this?

ttrei replied:
see commit message: https://github.com/NixOS/nixpkgs/pull/339921/commits/03266aa56ad8bcc9d35c6f9d049af6d2e9d9db4a

aarnphm commented on maintainers/maintainer-list.nix:
```diff
@@ -21429,6 +21429,12 @@
+    github = "ttrei";
+    githubId = 27609929;
+    name = "Reinis Taukulis";
+  };
```
```suggestion
  };
  aarnphm = {
    email = "contact@aarnphm.xyz";
    github = "aarnphm";
    githubId = 29749331;
    name = "Aaron Pham";
  };
```

aarnphm commented on pkgs/applications/editors/vim/plugins/avante-nvim/default.nix:
```diff
@@ -0,0 +1,71 @@
+    description = "Neovim plugin designed to emulate the behaviour of the Cursor AI IDE";
+    homepage = "https://github.com/yetone/avante.nvim";
+    license = licenses.asl20;
+    maintainers = with lib.maintainers; [ ttrei ];
```
```suggestion
    maintainers = with lib.maintainers; [ ttrei aarnphm ];
```

ttrei replied:
committed

aarnphm reviewed (commented):
few tiny things, add myself as maintainer for help

ttrei committed:
maintainers: add arrnphm

GaetanLepage commented on pkgs/applications/editors/vim/plugins/avante-nvim/default.nix:
```diff
@@ -0,0 +1,74 @@
+}:
+
+let
+  version = "2024-09-15";
+
+  src = fetchFromGitHub {
+    owner = "yetone";
+    repo = "avante.nvim";
+    rev = "f9520c4fdfed08e9cc609d6cd319b358e4ea33a5";
+    hash = "sha256-8zTDGPnhNI2rQA0uJc8gQRj4JCyg+IkO/D3oHYy4f9U=";
+  };
```
I would recommend to follow the same approach as in #341927 and use the upstream releases (0.0.2 currently).
Then, you will be able to add `passthru.updateScript = nix-update-script { };` which should make our lives easier.

ttrei replied:
I don't like moving to upstream releases because
- 0.0.2 is broken (see comments above)
- I don't want to impose a release schedule on upstream
- Almost all other plugins specify a commit, not a release tag

Regarding the update script, i did some limited testing, it seemed to not update Cargo.lock. Maybe user error.
The only vim plugin that has updateScript is https://github.com/NixOS/nixpkgs/blob/f9e21602e2b6954ee32056c9525dc150babf5dc5/pkgs/applications/editors/vim/plugins/vim-clap/default.nix#L70
hmm, is it used just to more easily populate outputHashes?

GaetanLepage replied:
> I don't like moving to upstream releases because
> 
>     * 0.0.2 is broken (see comments above)
> 
>     * I don't want to impose a release schedule on upstream
> 
>     * Almost all other plugins specify a commit, not a release tag

Fair enough !

GaetanLepage commented on pkgs/applications/editors/vim/plugins/neovim-require-check-hook.sh:
```diff
@@ -10,7 +10,7 @@ neovimRequireCheckHook () {
         export HOME="$TMPDIR"
         @nvimBinary@ -es --headless -n -u NONE -i NONE --clean -V1 \
-            --cmd "set rtp+=$out,${dependencies/ /,}" \
+            --cmd "set rtp+=$out,${dependencies// /,}" \
```
This patch should definitely get its own PR.
cc @teto

GaetanLepage replied:
Also, I am afraid that it will need to target staging as the number of rebuilds is significant.

ttrei replied:
I will extract into separate pr

ttrei replied:
https://github.com/NixOS/nixpkgs/pull/342240
will disable the check hook while the fix is not merged

GaetanLepage commented:
Result of `nixpkgs-review pr 339921` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>vimPlugins.avante-nvim</li>
    <li>vimPluginsUpdater</li>
  </ul>
</details>

GaetanLepage reviewed (approved):
LGTM ! Thanks @ttrei

GaetanLepage commented:
Result of `nixpkgs-review pr 339921` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>vimPlugins.avante-nvim</li>
    <li>vimPluginsUpdater</li>
  </ul>
</details>

GaetanLepage commented:
Result of `nixpkgs-review pr 339921` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>vimPlugins.avante-nvim</li>
    <li>vimPluginsUpdater</li>
  </ul>
</details>
